### PR TITLE
haskellPackages.tensorflow-mnist-input-data: fetch

### PR DESCRIPTION
**Pull request by teto**
###### Motivation for this change
Prefetch data so that the package doesn't try to connect to the internet.
See https://github.com/NixOS/nixpkgs/issues/46029

`nix-build -A haskellPackages.tensorflow-mnist ~/nixpkgs` still fails though because of 
```
Setup: Encountered missing dependencies:
tensorflow-mnist-input-data -any
```
definition of the package tensorflow-mnist seems
```
       broken = true;
     }) {tensorflow-mnist-input-data = null;};
```
@peti git blame says that code is generated. Anyway to not set tensorflow-mnist-input-data to null ?

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-tensorflow.nix [hunk 1]
```diff
@@ -62,6 +62,30 @@ in
     { mkDerivation, base, bytestring, Cabal, cryptonite, directory
     , filepath, HTTP, network-uri, stdenv
     }:
+
+    let
+      urlPrefix = "http://yann.lecun.com/exdb/mnist/";
+
+      # File names relative to 'urlPrefix' and their sha256.
+      fileInfos = [
```
Comment (basvandijk) [inline pkgs/development/haskell-modules/configuration-tensorflow.nix]: I think using an attribute set here which maps file names to their hashes is a bit clearer and safer to use.

File: pkgs/development/haskell-modules/configuration-tensorflow.nix [hunk 2]
```diff
@@ -71,6 +95,9 @@ in
         base bytestring Cabal cryptonite directory filepath HTTP
         network-uri
       ];
+      preConfigure = pkgs.lib.strings.concatStringsSep "\n" (
+          map (x: "cp ${x} data/$(stripHash ${x})") downloads
```
Comment (basvandijk) [inline pkgs/development/haskell-modules/configuration-tensorflow.nix]: Maybe use symbolic links instead of copying the files to prevent duplication.

**Review by basvandijk — CHANGES_REQUESTED** [review]
Thanks for this fix. I should have tested building in a sandbox before committing the tensorflow packages.

**Comment by basvandijk** [comment]
I've no idea where the `broken = true` and the `tensorflow-mnist-input-data = null` override in:

```
"tensorflow-mnist" = 
  ...
       broken = true;
     }) {tensorflow-mnist-input-data = null;};
```

are coming from. `tensorflow-mnist` is not mentioned in [configuration-hackage2nix.yaml](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/haskell-modules/configuration-hackage2nix.yaml) nor in [cabal2nix](https://github.com/NixOS/cabal2nix).

**Comment by basvandijk** [comment]
After reading [the source](https://github.com/NixOS/cabal2nix/blob/master/src/Distribution/Nixpkgs/Haskell/FromCabal.hs#L113
) I understand that packages are marked `broken` when they have missing dependencies.  The `tensorflow-mnist-input-data` dependency of `tensorflow-mnist` is probably considered missing since it's not published on Hackage.

The best solution is to have an override for `tensorflow-mnist` where we get rid of the `broken = True` flag and point `tensorflow-mnist-input-data` to the actual package.

We should also report an issue at https://github.com/tensorflow/haskell/issues to publish `tensorflow-mnist-input-data` to Hackage.

**Comment by basvandijk** [comment]
I've now made these changes in this PR. I also reverted my `libtensorflow: 1.9 -> 1.10` upgrade since that broke the Haskell tensorflow packages and `libtensorflow` is only used by the Haskell packages so this revert doesn't break anything else.

I also removed the tensorflow packages from `dont-distribute-packages` so they will be build on Hydra again after the next `hackage2nix` run.

@GrahamcOfBorg build haskellPackages.tensorflow-mnist

**Review by basvandijk — APPROVED** [review]

**Comment by teto** [comment]
wow that's super cool !! thanks.

**Pull request merged**